cmd/nanotube/process.go

```diff
@@ -28,47 +29,57 @@ func Process(queue <-chan string, rules rules.Rules, workerPoolSize uint16, shou
 	return done
 }
 
-func worker(wg *sync.WaitGroup, queue <-chan string, rules rules.Rules, shouldValidate bool, shouldLog bool, lg *zap.Logger, metrics *metrics.Prom) {
+func worker(wg *sync.WaitGroup, queue <-chan string, rules rules.Rules, rewrites rewrites.Rewrites, shouldValidate bool, shouldLog bool, lg *zap.Logger, metrics *metrics.Prom) {
 	defer wg.Done()
 	for j := range queue {
-		proc(&j, rules, shouldValidate, shouldLog, lg, metrics)
+		proc(&j, rules, rewrites, shouldValidate, shouldLog, lg, metrics)
 	}
 }
 
-func proc(s *string, rules rules.Rules, shouldNormalize bool, shouldLog bool, lg *zap.Logger, metrics *metrics.Prom) {
-	r, err := rec.ParseRec(*s, shouldNormalize, shouldLog, time.Now, lg)
-	if err != nil {
-		lg.Info("Error parsing incoming record", zap.String("record", *s), zap.Error(err))
-		metrics.ErrorRecs.Inc()
-		return
-	}
-
+func pushRec(rec *rec.Rec, rules rules.Rules, lg *zap.Logger, metrics *metrics.Prom) {
 	pushedTo := make(map[*target.Cluster]struct{})
 
 	for _, rl := range rules {
-		matchedRules := false
-		for _, re := range rl.CompiledRE {
-			if re.MatchString(r.Path) {
-				matchedRules = true
-				for _, cl := range rl.Targets {
-					if _, pushedBefore := pushedTo[cl]; pushedBefore {
-						continue
-					}
-					err = cl.Push(r, metrics)
-					if err != nil {
-						lg.Error("push to cluster failed",
-							zap.Error(err),
-							zap.String("cluster", cl.Name),
-							zap.String("record", *r.Serialize()))
-					}
-					pushedTo[cl] = struct{}{}
+		matchedRule := rl.Match(rec)
+		if matchedRule {
+			for _, cl := range rl.Targets {
+				if _, pushedBefore := pushedTo[cl]; pushedBefore {
+					continue
+				}
+				err := cl.Push(rec, metrics)
+				if err != nil {
+					lg.Error("push to cluster failed",
+						zap.Error(err),
+						zap.String("cluster", cl.Name),
+						zap.String("record", *rec.Serialize()))
 				}
-				// if one regex in the group matches, move to next rule
-				break
+				pushedTo[cl] = struct{}{}
 			}
 		}
-		if matchedRules && !rl.Continue {
+
+		if matchedRule && !rl.Continue {
 			break
 		}
 	}
 }
+
+func proc(s *string, rules rules.Rules, rewrites rewrites.Rewrites, shouldNormalize bool, shouldLog bool, lg *zap.Logger, metrics *metrics.Prom) {
+	r, err := rec.ParseRec(*s, shouldNormalize, shouldLog, time.Now, lg)
+	if err != nil {
+		lg.Info("Error parsing incoming record", zap.String("record", *s), zap.Error(err))
+		metrics.ErrorRecs.Inc()
+		return
+	}
+
+	recs, err := rewrites.RewriteMetric(r)
+	if err != nil {
+		//even if we get error here, we should still receive back original record
```

Comment:
To make it more clear you may consider returning additional records from the `RewriteMetric` func. What do you think?

Reply:
TBH, there don't seem be any case where we can return error from RewriteMetric.

Reply:
Let me clarify. I mean: instead of returning all records from `RewriteMetric`, you may consider returning only additional records that are results of rewrite. The records that are passed are changed in-place. This way, error handling semantics can be made more clear. But this is debatable, if you don't feel like redoing this, let's leave as-is.

Reply:
I removed unused error.